File: BleModule/src/main/java/com/topdon/commons/poster/MethodInfo.java
package com.topdon.commons.poster;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.lang.reflect.Method;

public class MethodInfo {
    @NonNull
    private String name;
    @Nullable
    private Parameter[] parameters;
    @NonNull
    private String tag;

    public MethodInfo(@NonNull String name, @Nullable Parameter... parameters) {
        this(name, name, parameters);
    }

    public MethodInfo(@NonNull String name, @NonNull String tag, @Nullable Parameter... parameters) {
        this.name = name;
        this.tag = tag;
        this.parameters = parameters;
    }

    public MethodInfo(@NonNull String name, @Nullable Class<?>[] parameterTypes) {
        this(name, name, parameterTypes);
    }

    public MethodInfo(@NonNull String name, @NonNull String tag, @Nullable Class<?>[] parameterTypes) {
        this(name, tag, toParameters(parameterTypes));
    }

    public static MethodInfo valueOf(@NonNull Method method) {
        Tag annotation = method.getAnnotation(Tag.class);
        return new MethodInfo(method.getName(), annotation == null ? method.getName() : annotation.value(),
                method.getParameterTypes());
    }

    private static Parameter[] toParameters(Class<?>[] parameterTypes) {
        Parameter[] parameters = null;
        if (parameterTypes != null) {
            parameters = new Parameter[parameterTypes.length];
            for (int i = 0; i < parameterTypes.length; i++) {
                parameters[i] = new Parameter(parameterTypes[i], null);
            }
        }
        return parameters;
    }

    @NonNull
    public String getName() {
        return name;
    }

    public void setName(@NonNull String name) {
        this.name = name;
    }

    @NonNull
    public String getTag() {
        return tag;
    }

    public void setTag(@NonNull String tag) {
        this.tag = tag;
    }

    @Nullable
    public Parameter[] getParameters() {
        return parameters;
    }

    public void setParameters(@Nullable Parameter[] parameters) {
        this.parameters = parameters;
    }

    @Nullable
    public Class<?>[] getParameterTypes() {
        if (parameters == null) {
            return null;
        } else {
            Class<?>[] types = new Class[parameters.length];
            for (int i = 0; i < parameters.length; i++) {
                types[i] = parameters[i].type;
            }
            return types;
        }
    }

    @Nullable
    public Object[] getParameterValues() {
        if (parameters == null) {
            return null;
        } else {
            Object[] values = new Class[parameters.length];
            for (int i = 0; i < parameters.length; i++) {
                values[i] = parameters[i].value;
            }
            return values;
        }
    }

    public static class Parameter {
        @Nullable
        private Object value;
        @NonNull
        private Class<?> type;

        public Parameter(@NonNull Class<?> type, @Nullable Object value) {
            this.type = type;
            this.value = value;
        }

        @Nullable
        public Object getValue() {
            return value;
        }

        public void setValue(@Nullable Object value) {
            this.value = value;
        }

        @NonNull
        public Class<?> getType() {
            return type;
        }

        public void setType(@NonNull Class<?> type) {
            this.type = type;
        }
    }
}

File: ble-core/src/main/java/com/mpdc4gsr/commons/poster/MethodInfo.kt
package com.mpdc4gsr.commons.poster

import java.lang.reflect.Method

class MethodInfo(var name: String, var tag: String, vararg parameters: Parameter?) {
    var parameters: Array<Parameter?>?

    constructor(name: String, vararg parameters: Parameter?) : this(name, name, *parameters)

    init {
        this.parameters = arrayOf(*parameters)
    }

    constructor(name: String, parameterTypes: Array<Class<*>?>?) : this(name, name, parameterTypes)

    constructor(name: String, tag: String, parameterTypes: Array<Class<*>?>?) : this(
        name,
        tag,
        *toParameters(parameterTypes)!!
    )

    val parameterTypes: Array<Class<*>?>?
        get() {
            val params = parameters
            if (params == null) {
                return null
            } else {
                val types = arrayOfNulls<Class<*>>(params.size)
                for (i in params.indices) {
                    types[i] = params[i]!!.type
                }
                return types
            }
        }

    val parameterValues: Array<Any?>?
        get() {
            val params = parameters
            if (params == null) {
                return null
            } else {
                val values: Array<Any?> = arrayOfNulls<Any>(params.size)
                for (i in params.indices) {
                    values[i] = params[i]!!.value
                }
                return values
            }
        }

    /**
     * Get the method name
     */
    fun getName(): String {
        return name
    }

    /**
     * Get the tag
     */
    fun getTag(): String {
        return tag
    }

    /**
     * Get the parameters array
     */
    fun getParameters(): Array<Parameter?>? {
        return parameters
    }

    /**
     * Get parameter types
     */
    fun getParameterTypes(): Array<Class<*>?>? {
        return parameterTypes
    }

    class Parameter(var type: Class<*>, var value: Any?) {
        /**
         * Get the parameter type
         */
        fun getType(): Class<*> {
            return type
        }

        /**
         * Get the parameter value
         */
        fun getValue(): Any? {
            return value
        }
    }
    companion object {
        fun valueOf(method: Method): MethodInfo {
            val annotation = method.getAnnotation<Tag?>(Tag::class.java)
            return MethodInfo(
                method.name, if (annotation == null) method.name else annotation.value,
                method.parameterTypes
            )
        }

        private fun toParameters(parameterTypes: Array<Class<*>?>?): Array<Parameter?>? {
            var parameters: Array<Parameter?>? = null
            if (parameterTypes != null) {
                parameters = arrayOfNulls<Parameter>(parameterTypes.size)
                for (i in parameterTypes.indices) {
                    parameters[i] = MethodInfo.Parameter(parameterTypes[i]!!, null)
                }
            }
            return parameters
        }
    }
}
